File: scripts/generate_cyber_pdfs.py
import json
from pathlib import Path
from textwrap import wrap


ROOT = Path(__file__).resolve().parent.parent
SUBMISSIONS_DIR = ROOT / "data" / "submissions"

PDF_SPECS = [
    {
        "file_name": "company-description.pdf",
        "file_type": "pdf",
        "category": "company_description",
        "description": "Company description and cyber exposure overview",
        "title": "Company Description",
    },
    {
        "file_name": "overall-status.pdf",
        "file_type": "pdf",
        "category": "overall_status",
        "description": "Overall cyber underwriting status summary",
        "title": "Overall Status",
    },
]


def main():
    for folder_path in sorted(SUBMISSIONS_DIR.iterdir()):
        metadata_path = folder_path / "metadata.json"
        if not metadata_path.exists():
            continue

        metadata = json.loads(metadata_path.read_text(encoding="utf-8"))

        for spec in PDF_SPECS:
            created_at = metadata.get("updated_at") or metadata.get("file_created_at")
            lines = (
                build_company_description(metadata)
                if spec["category"] == "company_description"
                else build_overall_status(metadata)
            )
            (folder_path / spec["file_name"]).write_bytes(create_pdf(spec["title"], lines))

            document_metadata = {
                "file_name": spec["file_name"],
                "file_type": spec["file_type"],
                "category": spec["category"],
                "file_created_at": created_at,
                "received_at": metadata.get("received_at"),
                "description": spec["description"],
            }

            existing = next(
                (
                    document
                    for document in metadata.get("documents", [])
                    if document.get("file_name") == spec["file_name"]
                ),
                None,
            )
            if existing:
                existing.update(document_metadata)
            else:
                metadata.setdefault("documents", []).append(document_metadata)

        metadata_path.write_text(json.dumps(metadata, indent=2) + "\n", encoding="utf-8")


def build_company_description(metadata):
    applicant = metadata.get("applicant", {})
    coverage = metadata.get("coverage", {})

    return [
        metadata.get("title"),
        "",
        f"Insured: {applicant.get('insured_name')}",
        f"Industry: {applicant.get('industry')}",
        f"Location: {applicant.get('location')}",
        f"Annual Revenue: {format_currency(applicant.get('annual_revenue'))}",
        f"Employees: {applicant.get('employee_count')}",
        f"Records: {format_number(applicant.get('records_count'))}",
        "",
        "Technology Profile",
        applicant.get("technology_profile"),
        "",
        "Cyber Coverage Requested",
        *[f"- {line}" for line in coverage.get("lines_requested", [])],
        f"Effective Date: {coverage.get('requested_effective_date')}",
        f"Retention Requested: {coverage.get('retention_requested')}",
    ]


def build_overall_status(metadata):
    controls = metadata.get("security_controls", {})

    return [
        metadata.get("title"),
        "",
        f"Submission ID: {metadata.get('id')}",
        f"Line of Business: {metadata.get('line_of_business')}",
        f"Status: {metadata.get('status')}",
        f"File Created At: {metadata.get('file_created_at')}",
        f"Received At: {metadata.get('received_at')}",
        f"Updated At: {metadata.get('updated_at')}",
        "",
        "Security Controls",
        *[f"- {labelize(key)}: {value}" for key, value in controls.items()],
        "",
        "Risk Flags",
        *[f"- {flag}" for flag in metadata.get("risk_flags", [])],
        "",
        "Open Questions",
        *[f"- {question}" for question in metadata.get("open_questions", [])],
    ]


def create_pdf(title, lines):
    escaped_lines = []
    for line in lines:
        escaped_lines.extend(wrap(str(line or ""), width=92) or [""])

    text_commands = "\n".join(
        [
            "BT",
            "/F1 18 Tf",
            "72 740 Td",
            f"({escape_pdf_text(title)}) Tj",
            "/F1 10 Tf",
            "0 -28 Td",
            *[f"({escape_pdf_text(line)}) Tj 0 -15 Td" for line in escaped_lines],
            "ET",
        ]
    )

    objects = [
        "<< /Type /Catalog /Pages 2 0 R >>",
        "<< /Type /Pages /Kids [3 0 R] /Count 1 >>",
        "<< /Type /Page /Parent 2 0 R /MediaBox [0 0 612 792] /Resources << /Font << /F1 4 0 R >> >> /Contents 5 0 R >>",
        "<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica >>",
        f"<< /Length {len(text_commands.encode('utf-8'))} >>\nstream\n{text_commands}\nendstream",
    ]

    pdf = "%PDF-1.4\n"
    offsets = [0]
    for index, pdf_object in enumerate(objects, start=1):
        offsets.append(len(pdf.encode("utf-8")))
        pdf += f"{index} 0 obj\n{pdf_object}\nendobj\n"

    xref_offset = len(pdf.encode("utf-8"))
    pdf += f"xref\n0 {len(objects) + 1}\n"
    pdf += "0000000000 65535 f \n"
    for offset in offsets[1:]:
        pdf += f"{offset:010d} 00000 n \n"
    pdf += f"trailer\n<< /Size {len(objects) + 1} /Root 1 0 R >>\nstartxref\n{xref_offset}\n%%EOF\n"

    return pdf.encode("utf-8")


def escape_pdf_text(value):
    return str(value).replace("\\", "\\\\").replace("(", "\\(").replace(")", "\\)")


def labelize(value):
    return " ".join(part.capitalize() for part in str(value).split("_"))


def format_currency(value):
    return f"${value:,.0f}" if isinstance(value, (int, float)) else "TBD"


def format_number(value):
    return f"{value:,}" if isinstance(value, (int, float)) else "TBD"


if __name__ == "__main__":
    main()

File: scripts/generate_cyber_pdfs.js
const fs = require("fs");
const path = require("path");

const ROOT = path.join(__dirname, "..");
const SUBMISSIONS_DIR = path.join(ROOT, "data", "submissions");

const pdfSpecs = [
  {
    file_name: "company-description.pdf",
    file_type: "pdf",
    category: "company_description",
    description: "Company description and cyber exposure overview",
    title: "Company Description"
  },
  {
    file_name: "overall-status.pdf",
    file_type: "pdf",
    category: "overall_status",
    description: "Overall cyber underwriting status summary",
    title: "Overall Status"
  }
];

for (const folderName of fs.readdirSync(SUBMISSIONS_DIR).sort()) {
  const folderPath = path.join(SUBMISSIONS_DIR, folderName);
  const metadataPath = path.join(folderPath, "metadata.json");

  if (!fs.existsSync(metadataPath)) {
    continue;
  }

  const metadata = JSON.parse(fs.readFileSync(metadataPath, "utf8"));

  for (const spec of pdfSpecs) {
    const createdAt = metadata.updated_at || metadata.file_created_at;
    const lines =
      spec.category === "company_description"
        ? buildCompanyDescription(metadata)
        : buildOverallStatus(metadata);

    fs.writeFileSync(path.join(folderPath, spec.file_name), createPdf(spec.title, lines));

    const existing = metadata.documents.find((document) => document.file_name === spec.file_name);
    const documentMetadata = {
      file_name: spec.file_name,
      file_type: spec.file_type,
      category: spec.category,
      file_created_at: createdAt,
      received_at: metadata.received_at,
      description: spec.description
    };

    if (existing) {
      Object.assign(existing, documentMetadata);
    } else {
      metadata.documents.push(documentMetadata);
    }
  }

  fs.writeFileSync(metadataPath, `${JSON.stringify(metadata, null, 2)}\n`);
}

function buildCompanyDescription(metadata) {
  const applicant = metadata.applicant || {};
  const coverage = metadata.coverage || {};

  return [
    metadata.title,
    "",
    `Insured: ${applicant.insured_name}`,
    `Industry: ${applicant.industry}`,
    `Location: ${applicant.location}`,
    `Annual Revenue: ${formatCurrency(applicant.annual_revenue)}`,
    `Employees: ${applicant.employee_count}`,
    `Records: ${formatNumber(applicant.records_count)}`,
    "",
    "Technology Profile",
    applicant.technology_profile,
    "",
    "Cyber Coverage Requested",
    ...(coverage.lines_requested || []).map((line) => `- ${line}`),
    `Effective Date: ${coverage.requested_effective_date}`,
    `Retention Requested: ${coverage.retention_requested}`
  ];
}

function buildOverallStatus(metadata) {
  const controls = metadata.security_controls || {};

  return [
    metadata.title,
    "",
    `Submission ID: ${metadata.id}`,
    `Line of Business: ${metadata.line_of_business}`,
    `Status: ${metadata.status}`,
    `File Created At: ${metadata.file_created_at}`,
    `Received At: ${metadata.received_at}`,
    `Updated At: ${metadata.updated_at}`,
    "",
    "Security Controls",
    ...Object.entries(controls).map(([key, value]) => `- ${labelize(key)}: ${value}`),
    "",
    "Risk Flags",
    ...(metadata.risk_flags || []).map((flag) => `- ${flag}`),
    "",
    "Open Questions",
    ...(metadata.open_questions || []).map((question) => `- ${question}`)
  ];
}

function createPdf(title, lines) {
  const escapedLines = lines.flatMap((line) => wrapLine(String(line || ""), 92));
  const textCommands = [
    "BT",
    "/F1 18 Tf",
    "72 740 Td",
    `(${escapePdfText(title)}) Tj`,
    "/F1 10 Tf",
    "0 -28 Td",
    ...escapedLines.map((line) => `(${escapePdfText(line)}) Tj 0 -15 Td`),
    "ET"
  ].join("\n");

  const objects = [
    "<< /Type /Catalog /Pages 2 0 R >>",
    "<< /Type /Pages /Kids [3 0 R] /Count 1 >>",
    "<< /Type /Page /Parent 2 0 R /MediaBox [0 0 612 792] /Resources << /Font << /F1 4 0 R >> >> /Contents 5 0 R >>",
    "<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica >>",
    `<< /Length ${Buffer.byteLength(textCommands)} >>\nstream\n${textCommands}\nendstream`
  ];

  let pdf = "%PDF-1.4\n";
  const offsets = [0];

  objects.forEach((object, index) => {
    offsets.push(Buffer.byteLength(pdf));
    pdf += `${index + 1} 0 obj\n${object}\nendobj\n`;
  });

  const xrefOffset = Buffer.byteLength(pdf);
  pdf += `xref\n0 ${objects.length + 1}\n`;
  pdf += "0000000000 65535 f \n";
  offsets.slice(1).forEach((offset) => {
    pdf += `${String(offset).padStart(10, "0")} 00000 n \n`;
  });
  pdf += `trailer\n<< /Size ${objects.length + 1} /Root 1 0 R >>\nstartxref\n${xrefOffset}\n%%EOF\n`;

  return Buffer.from(pdf, "utf8");
}

function wrapLine(line, maxLength) {
  if (line.length <= maxLength) {
    return [line];
  }

  const words = line.split(" ");
  const wrapped = [];
  let current = "";

  for (const word of words) {
    if (`${current} ${word}`.trim().length > maxLength) {
      wrapped.push(current);
      current = word;
    } else {
      current = `${current} ${word}`.trim();
    }
  }

  if (current) {
    wrapped.push(current);
  }

  return wrapped;
}

function escapePdfText(value) {
  return value.replace(/\\/g, "\\\\").replace(/\(/g, "\\(").replace(/\)/g, "\\)");
}

function labelize(value) {
  return value
    .split("_")
    .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
    .join(" ");
}

function formatCurrency(value) {
  return typeof value === "number" ? `$${value.toLocaleString("en-US")}` : "TBD";
}

function formatNumber(value) {
  return typeof value === "number" ? value.toLocaleString("en-US") : "TBD";
}
